### PR TITLE
Add CI job for testing vsphere IPI in MAO

### PR DESCRIPTION
**Pull request by alexander-demicev**
Add CI job for testing vsphere IPI

### CODE REVIEW

File: ci-operator/jobs/openshift/machine-api-operator/openshift-machine-api-operator-master-presubmits.yaml
Comment (enxebre) [inline ci-operator/jobs/openshift/machine-api-operator/openshift-machine-api-operator-master-presubmits.yaml:77]: The test we want to run is our machine e2e test suite against an existing IPI cluster.
We need to add to the config https://github.com/openshift/release/blob/037152beab98b4014f1ce7811655e53f477a07f3/ci-operator/config/openshift/machine-api-operator/openshift-machine-api-operator-master.yaml#L55-L60 something the likes of:

```- as: e2e-vsphere-operator
  commands: JUNIT_DIR=${ARTIFACT_DIR} make test-e2e
  openshift_installer_custom_test_image:
    cluster_profile: vsphere
    from: stable:cluster-api-actuator-pkg
    nested_virt_image: ""
```
Then run `make jobs` will generate the jobs for you.

We want this job to be `optional: true`

See for more details https://github.com/openshift/release/blob/master/ci-operator/README.md

Comment (enxebre) [reply ci-operator/jobs/openshift/machine-api-operator/openshift-machine-api-operator-master-presubmits.yaml:77]: Then from the mao repo we should be able to run `/test e2e-vsphere-operator`
Is the installer already running e2e against IPI?

Comment (enxebre) [reply ci-operator/jobs/openshift/machine-api-operator/openshift-machine-api-operator-master-presubmits.yaml:77]: Seems  we are using steps now
https://github.com/openshift/release/blob/master/ci-operator/config/openshift/installer/openshift-installer-master.yaml#L201-L204

So we need to find the analogous for `openshift_installer_custom_test_image` with steps. See
https://steps.svc.ci.openshift.org/

Comment (enxebre) [reply ci-operator/jobs/openshift/machine-api-operator/openshift-machine-api-operator-master-presubmits.yaml:77]: See also https://steps.svc.ci.openshift.org/help/examples

I think we should be able to have something like

```
- as: e2e-vsphere
  steps:
    cluster_profile: vsphere
    workflow: origin-e2e-vsphere
    test:
     From: stable:cluster-api-actuator-pkg
     commands:
        - JUNIT_DIR=${ARTIFACT_DIR} make test-e2e
```

File: ci-operator/jobs/openshift/machine-api-operator/openshift-machine-api-operator-master-presubmits.yaml
Comment (enxebre) [inline ci-operator/jobs/openshift/machine-api-operator/openshift-machine-api-operator-master-presubmits.yaml:846]: This needs to be e2e-vsphere-operator like we do for any other provider running our machine API suite.

We use `e2e-providerX` for the origin suite over the installer

File: ci-operator/config/openshift/machine-api-operator/openshift-machine-api-operator-master.yaml
Comment (enxebre) [inline ci-operator/config/openshift/machine-api-operator/openshift-machine-api-operator-master.yaml:92]: https://github.com/openshift/release/pull/8487/files#r413675530

**Comment by enxebre** [comment]
I'm not sure this is still possible. But in the past you could test and run your job locally 
e.g.
```
 ./ci-operator --config ~/go/src/github.com/openshift/release/ci-operator/config/openshift/cluster-api-provider-gcp/openshift-cluster-api-provider-gcp-master.yaml --git-ref openshift/cluster-api-provider-gcp@master
```
https://github.com/openshift/release/tree/master/ci-operator#testing-a-template

File: ci-operator/config/openshift/machine-api-operator/openshift-machine-api-operator-master.yaml
Comment (alexander-demichev) [reply ci-operator/config/openshift/machine-api-operator/openshift-machine-api-operator-master.yaml]: fixed

**Comment by alexander-demichev** [comment]
Will CI verify that PR works?

**Comment by enxebre** [comment]
>error: could not run steps: step e2e-vsphere-operator failed: failed to acquire lease: resources not found
?
/retest

File: ci-operator/jobs/openshift/machine-api-operator/openshift-machine-api-operator-master-presubmits.yaml
Comment (jcpowermac) [inline ci-operator/jobs/openshift/machine-api-operator/openshift-machine-api-operator-master-presubmits.yaml:799]: @alexander-demichev the cluster needs to be `vsphere`

Comment (jcpowermac) [reply ci-operator/jobs/openshift/machine-api-operator/openshift-machine-api-operator-master-presubmits.yaml]: See:
https://github.com/openshift/release/blob/master/core-services/sanitize-prow-jobs/_config.yaml#L3217-L3268
Make sure you run `make jobs`

**Review by michaelgugino — CHANGES_REQUESTED** [review]
/hold

Need detailed documentation about how this environment is set up so we have a reference we can point users to when their environment fails and ours didn't.

**Comment by jcpowermac** [comment]
> Need detailed documentation about how this environment is set up so we have a reference we can point users to when their environment fails and ours didn't.

@michaelgugino every vSphere enviornment is a snowflake. I am not sure what this will accomplish. I would rather see a signal of MAO functionality then delaying not knowning any status.

**Comment by michaelgugino** [comment]
> > Need detailed documentation about how this environment is set up so we have a reference we can point users to when their environment fails and ours didn't.
> 
> @michaelgugino every vSphere enviornment is a snowflake. I am not sure what this will accomplish. I would rather see a signal of MAO functionality then delaying not knowning any status.

Indeed, every user cluster is a snowflake.  That's why we need to document what OUR setup is so when support cases are opened we can compare our settings with theirs.  Otherwise we have no ideas what problems they're running into and whether or not their configuration is actually supported.

**Comment by jcpowermac** [comment]
> > > Need detailed documentation about how this environment is set up so we have a reference we can point users to when their environment fails and ours didn't.
> > 
> > 
> > @michaelgugino every vSphere enviornment is a snowflake. I am not sure what this will accomplish. I would rather see a signal of MAO functionality then delaying not knowning any status.
> 
> Indeed, every user cluster is a snowflake. That's why we need to document what OUR setup is so when support cases are opened we can compare our settings with theirs. Otherwise we have no ideas what problems they're running into and whether or not their configuration is actually supported.

Fortunately we are using VMware Cloud on AWS. VMware provides their configuration for vSphere and we just consume it. See:
https://docs.vmware.com/en/VMware-Cloud-on-AWS/index.html

**Comment by michaelgugino** [comment]
> > > > Need detailed documentation about how this environment is set up so we have a reference we can point users to when their environment fails and ours didn't.
> > > 
> > > 
> > > @michaelgugino every vSphere enviornment is a snowflake. I am not sure what this will accomplish. I would rather see a signal of MAO functionality then delaying not knowning any status.
> > 
> > 
> > Indeed, every user cluster is a snowflake. That's why we need to document what OUR setup is so when support cases are opened we can compare our settings with theirs. Otherwise we have no ideas what problems they're running into and whether or not their configuration is actually supported.
> 
> Fortunately we are using VMware Cloud on AWS. VMware provides their configuration for vSphere and we just consume it. See:
> https://docs.vmware.com/en/VMware-Cloud-on-AWS/index.html

This documentation seems to detail how AWS is configured to support VMWare, rather than the settings of the VMWare components themselves.  What we need is a process to replicate this environment onto an on-premise deployment.  I don't find this documentation to be sufficient for that purpose, thus we won't have a way to compare a supported configuration with what a user has.  Such a document would allow us to provide a questionnaire to users so we can compare their environment to ours, and investigate any drift as the potential for misconfiguration or bugs.

**Comment by jcpowermac** [comment]
> > > > > Need detailed documentation about how this environment is set up so we have a reference we can point users to when their environment fails and ours didn't.
> > > > 
> > > > 
> > > > @michaelgugino every vSphere enviornment is a snowflake. I am not sure what this will accomplish. I would rather see a signal of MAO functionality then delaying not knowning any status.
> > > 
> > > 
> > > Indeed, every user cluster is a snowflake. That's why we need to document what OUR setup is so when support cases are opened we can compare our settings with theirs. Otherwise we have no ideas what problems they're running into and whether or not their configuration is actually supported.
> > 
> > 
> > Fortunately we are using VMware Cloud on AWS. VMware provides their configuration for vSphere and we just consume it. See:
> > https://docs.vmware.com/en/VMware-Cloud-on-AWS/index.html
> 
> This documentation seems to detail how AWS is configured to support VMWare, rather than the settings of the VMWare components themselves. What we need is a process to replicate this environment onto an on-premise deployment. I don't find this documentation to be sufficient for that purpose, thus we won't have a way to compare a supported configuration with what a user has. Such a document would allow us to provide a questionnaire to users so we can compare their environment to ours, and investigate any drift as the potential for misconfiguration or bugs.

Is there a story in Jira to create this documentation? @michaelgugino since you are blocking this PR are you owning it?

**Comment by michaelgugino** [comment]
> > > > > > Need detailed documentation about how this environment is set up so we have a reference we can point users to when their environment fails and ours didn't.
> > > > > 
> > > > > 
> > > > > @michaelgugino every vSphere enviornment is a snowflake. I am not sure what this will accomplish. I would rather see a signal of MAO functionality then delaying not knowning any status.
> > > > 
> > > > 
> > > > Indeed, every user cluster is a snowflake. That's why we need to document what OUR setup is so when support cases are opened we can compare our settings with theirs. Otherwise we have no ideas what problems they're running into and whether or not their configuration is actually supported.
> > > 
> > > 
> > > Fortunately we are using VMware Cloud on AWS. VMware provides their configuration for vSphere and we just consume it. See:
> > > https://docs.vmware.com/en/VMware-Cloud-on-AWS/index.html
> > 
> > 
> > This documentation seems to detail how AWS is configured to support VMWare, rather than the settings of the VMWare components themselves. What we need is a process to replicate this environment onto an on-premise deployment. I don't find this documentation to be sufficient for that purpose, thus we won't have a way to compare a supported configuration with what a user has. Such a document would allow us to provide a questionnaire to users so we can compare their environment to ours, and investigate any drift as the potential for misconfiguration or bugs.
> 
> Is there a story in Jira to create this documentation? @michaelgugino since you are blocking this PR are you owning it?

No, I don't own the documentation, I don't know if there's a Jira story or not.  However owns the VMWare environment should own it.

**Comment by jcpowermac** [comment]
> > > > > > > Need detailed documentation about how this environment is set up so we have a reference we can point users to when their environment fails and ours didn't.
> > > > > > 
> > > > > > 
> > > > > > @michaelgugino every vSphere enviornment is a snowflake. I am not sure what this will accomplish. I would rather see a signal of MAO functionality then delaying not knowning any status.
> > > > > 
> > > > > 
> > > > > Indeed, every user cluster is a snowflake. That's why we need to document what OUR setup is so when support cases are opened we can compare our settings with theirs. Otherwise we have no ideas what problems they're running into and whether or not their configuration is actually supported.
> > > > 
> > > > 
> > > > Fortunately we are using VMware Cloud on AWS. VMware provides their configuration for vSphere and we just consume it. See:
> > > > https://docs.vmware.com/en/VMware-Cloud-on-AWS/index.html
> > > 
> > > 
> > > This documentation seems to detail how AWS is configured to support VMWare, rather than the settings of the VMWare components themselves. What we need is a process to replicate this environment onto an on-premise deployment. I don't find this documentation to be sufficient for that purpose, thus we won't have a way to compare a supported configuration with what a user has. Such a document would allow us to provide a questionnaire to users so we can compare their environment to ours, and investigate any drift as the potential for misconfiguration or bugs.
> > 
> > 
> > Is there a story in Jira to create this documentation? @michaelgugino since you are blocking this PR are you owning it?
> 
> No, I don't own the documentation, I don't know if there's a Jira story or not. However owns the VMWare environment should own it.

Here is another doc: https://www.vmware.com/content/dam/digitalmarketing/vmware/en/pdf/partners/vmware-cloudtm-on-aws-technical-overview.pdf

If that isn't sufficent to your expectations you will need to reach out to DPP, they own the enivornment.

**Comment by alexander-demichev** [comment]
@michaelgugino I agree with you here but the problem you are describing is not only related to `machine-api`. Other openshift components are also tested in the way. I think you should raise this topic on a mailing list. Right now we are doing a blind development of machine-api for vSphere and this PR should be merged, every change we do for vsphere doesn't have e2e coverage.

**Comment by enxebre** [comment]
@alexander-demichev @jcpowermac this is still failing https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_release/8487/rehearse-8487-pull-ci-openshift-machine-api-operator-master-e2e-vsphere-operator/1303340886620377088

/retest

File: ci-operator/config/openshift/machine-api-operator/openshift-machine-api-operator-master.yaml
Comment (Danil-Grigorev) [inline ci-operator/config/openshift/machine-api-operator/openshift-machine-api-operator-master.yaml:129]: This isn't supposed to work already. It seems we are using multi-step vsphere job, and so you should run the tests similar to this example - https://github.com/openshift/release/blob/71ba7717f2ef5d0001bffe1185b800730230e9ac/ci-operator/config/openshift/cluster-image-registry-operator/openshift-cluster-image-registry-operator-master.yaml#L91-L103. You should also set the tests against `cluster-api-actuator-pkg` as I'm not sure if building a custom test image, pointing to other repos similar to what `openshift_installer_custom_test_image` provided is now supported. @jcpowermac Could you assist answering this question?

**Review by Danil-Grigorev — CHANGES_REQUESTED** [review]

**Comment by JoelSpeed** [comment]
> Here is another doc: https://www.vmware.com/content/dam/digitalmarketing/vmware/en/pdf/partners/vmware-cloudtm-on-> aws-technical-overview.pdf
>
> If that isn't sufficent to your expectations you will need to reach out to DPP, they own the enivornment.

I wanted to add my 2 cents to the discussion that was being had about how the environment is set up. I agree with @michaelgugino that we should have somewhere documented (IMO publicly) what configuration of vSphere we are testing against. Obviously every vSphere env is different but there are some broad feature sets that we can encapsulate (eg how is the network and storage configured, have we tested vMotion) into a document. This will help teams like ours to determine differences between our test and dev environments (which as far as I'm aware are all the same) and the customer environment.

The document link above (from the quoted text from @jcpowermac, thank you) does, to me at least, seem to answer a lot of the questions for what our standard test environment is. I don't use vSphere very much and am largely unfamiliar with it, but it seems to detail how the storage is configured and mentions the networking is based on NSX, it also talks briefly about how vMotion is used regularly within these environments.

It would be great to determine the important parts of this doc and decant them into some of our own documentation to summarise the vSphere configuration that we are testing against, perhaps with the help of some vSphere experts within the org who understand the different terms better.  

With all of that said, I don't feel the need for us to block our testing on this document being created. To me, these are tangential issues and my concerns about our code regressing for some subset of the vSphere clusters (the ones that match our test config) outweighs my desire to have this documented. I feel like having some signal for CI for vSphere, while nowhere near as good as the signal we have for AWS/GCP etc, is better than nothing. It covers the broad functionality of vSphere and should give us at least some confidence in our changes until we can improve the matrix of test environments for vSphere.

File: ci-operator/config/openshift/machine-api-operator/openshift-machine-api-operator-master.yaml
Comment (Danil-Grigorev) [inline ci-operator/config/openshift/machine-api-operator/openshift-machine-api-operator-master.yaml:140]: This is missing `workflow: openshift-e2e-vsphere` on the next line.

File: ci-operator/config/openshift/machine-api-operator/openshift-machine-api-operator-master.yaml
Comment (Danil-Grigorev) [inline ci-operator/config/openshift/machine-api-operator/openshift-machine-api-operator-master.yaml:133]: I don't think we need this line.

**Comment by JoelSpeed** [comment]
Looks like we made some progress on this, we are actually running the test suites now in the operator tests at least. Not getting the Junit breakdown at the moment though, we should make sure we have that before we merge

**Comment by JoelSpeed** [comment]
The test failures here are legitimate failures from tests that we have never tried out properly. The configuration seems to be correct at this point.
/lgtm

**Comment by Danil-Grigorev** [comment]
@JoelSpeed @alexander-demichev It is not correct, according to the thing we discussed around `e2e-operator` jobs yesterday. The state of the PR more confirms the other option, where we run all our e2e and origin tests in this job. We first need to implement the job template itself in https://github.com/openshift/release/tree/master/ci-operator/step-registry/ipi/vsphere, similar to https://github.com/openshift/release/blob/2e9568a67b694e6fe56c04022832566efd7e6259/ci-operator%2Fstep-registry%2Fipi%2Fgcp%2Fipi-gcp-workflow.yaml

\hold
\lgtm cancel

**Comment by JoelSpeed** [comment]
@Danil-Grigorev As far as I can tell this PR produces jobs which conform with what we have with other platforms. We are getting two status checks, with the correct names, running the correct tasks. 

This may not be up to date with the best practices and modern configuration options, but I would rather we get some CI set up so that we can validate any changes to vSphere than delay this further making sure the configuration is up to date. This can be migrated later.

I believe my LGTM is still valid and should persist
/lgtm

I also don't believe we should be holding for this reason and should migrate to the new format in a separate PR.

**Pull request merged**
